### PR TITLE
Add BBIC volume plugin

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

**Comment by tribal-tec** [comment]
retest this please

File: plugins/BBIC/BBICFile.cpp [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    levels_ = _volGroup.getGroup(BBIC_GROUP_LEVELS).getNumberObjects();
+
+    register_lzf();
+}
+
+std::array<size_t, 3> File::getBlockCount(const uint32_t level) const
+{
+    // Alternatively, this information could be read from the bbic level
+    // groups
+    return {{(size_t)std::ceil(float(width_ >> level) / blockSize_),
```
Comment (karjonas) [inline plugins/BBIC/BBICFile.cpp]:
```suggestion
    return {{static_cast<size_t>(std::ceil(float(width_ >> level) / blockSize_)),
```

File: plugins/BBIC/BBICLoader.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    auto volume = volumeModel.model->getModel().createBrickedVolume(
+        dim, spacing, brayns::DataType::UINT8);
+    volume->setDataRange({0, 255});
+    volumeModel.volumes.push_back(volume);
+
+    volumeModel.toUpload.push_back({});
+    for (uint32_t x = 0; x < blockCount[0]; ++x)
+        for (uint32_t y = 0; y < blockCount[1]; ++y)
+            for (uint32_t z = 0; z < blockCount[2]; ++z)
+                volumeModel.toUpload[lod].push_back({{x, y, z}});
```
Comment (karjonas) [inline plugins/BBIC/BBICLoader.cpp]:
```suggestion
                volumeModel.toUpload[lod].emplace_back(x, y, z);
```

File: plugins/BBIC/BBICLoader.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    const auto& block = volumeModel.toUpload[lod].back();
+    volumeModel.toUpload[lod].pop_back();
+
+    const brayns::Vector3ui region_lo(block[0] * blockSize,
+                                      block[1] * blockSize,
+                                      block[2] * blockSize);
+
+    const auto& data = file.getData(lod, block);
+    const brayns::Vector3ui voxelBox(blockSize);
+    volume->setBrick((void*)data.data(), region_lo, voxelBox);
```
Comment (karjonas) [inline plugins/BBIC/BBICLoader.cpp]:
```suggestion
    volume->setBrick(static_cast<void*>(data.data()), region_lo, voxelBox);
```

**Comment by tribal-tec** [comment]
retest this please

File: plugins/BBIC/VolumeModel.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        const brayns::Vector3ui voxelSize(blockSize);
+
+        while (keepUploading && !toUpload_.empty())
+        {
+            const auto& block = toUpload_.back();
+            toUpload_.pop_back();
+            const brayns::Vector3ui position(block[0] * blockSize,
+                                             block[1] * blockSize,
+                                             block[2] * blockSize);
+            volume->setBrick((void*)file->getData(lod, block).data(), position,
```
Comment (karjonas) [inline plugins/BBIC/VolumeModel.cpp]:
```suggestion
            volume->setBrick(static_cast<void*>(file->getData(lod, block).data()), position,
```

File: plugins/BBIC/VolumeModel.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        model->getModel().removeVolume(volume);
+
+    volume = volumes[lod];
+    model->getModel().addVolume(volume);
+
+    auto& toUpload_ = toUpload[lod];
+    if (toUpload_.empty())
+        return;
+
+    uploadThread = std::thread([this, &toUpload_] {
```
Comment (rdumusc) [inline plugins/BBIC/VolumeModel.cpp]: move to startUploadThread() function?

File: plugins/BBIC/VolumeModel.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+#include <brayns/common/types.h>
+
+#include <memory>
+#include <thread>
+
+#include "BBICFile.h"
+
+namespace bbic
+{
+struct VolumeModel
```
Comment (rdumusc) [inline plugins/BBIC/VolumeModel.h]: I find it suspicious that all the members public, this look more like a class to me

**Comment by tribal-tec** [comment]
Thanks for the comments. Looks better now. Was indeed some old code, hence it's better if it's pushed and more maintained.

**Comment by tribal-tec** [comment]
retest this please

**Comment by tribal-tec** [comment]
ping

**Comment by tribal-tec** [comment]
retest this please

**Pull request merged**